cloud/cloud_env_impl.cc

```diff
@@ -2007,41 +1986,67 @@ size_t CloudEnvImpl::TEST_NumScheduledJobs() const {
   return scheduler_->TEST_NumScheduledJobs();
 };
 
-Status CloudEnvImpl::ApplyLocalCloudManifestDelta(const std::string& local_dbname,
-    const std::string& new_cookie,
-    const CloudManifestDelta& delta) {
+Status CloudEnvImpl::ApplyCloudManifestDelta(const CloudManifestDelta& delta) {
+  cloud_manifest_->AddEpoch(delta.file_num, delta.epoch);
+  return Status::OK();
+}
+
+Status CloudEnvImpl::RollNewCookie(const std::string& local_dbname,
+                                   const std::string& cookie,
+                                   const CloudManifestDelta& delta) const {
+  auto newCloudManifest = cloud_manifest_->clone();
```

Comment:
why do we need to clone here? Are you trying to make this function idempotent so we can retry it if something is wrong while `OpenPresumptiveLeader` ? If so, why is it not necessary for `ApplyCloudManifestDelta(CMDelta)`?

Reply:
The goal was not to make it idempotent. The goal was to avoid it having any affect on the running database. Presumptive leader might fail to become an actual leader (due to somebody else winning the race) and has to go back to being a follower without a restart. We call `ApplyCloudManifestDelta()` only after we become a real leader, so that's the point at which we commit the new epoch to the database.